src/BlTools.MssqlFluentSqlWrapper/FluentSqlCommand.cs

```diff
@@ -0,0 +1,520 @@
+            {
+                try
+                {
+                    if (_command.Connection.State != ConnectionState.Open)
+                    {
+                        await _command.Connection.OpenAsync();
+                    }
+
+                    using (_command)
+                    {
+                        switch (execType)
+                        {
+                            case ExecType.NonQuery:
+                            {
+                                result.NonQueryResult = await _command.ExecuteNonQueryAsync();
+                                break;
+                            }
+                            case ExecType.Scalar:
+                            {
+                                result.ScalarResult = await _command.ExecuteScalarAsync();
+                                break;
+                            }
+                            case ExecType.Reader:
+                            {
+                                using var reader = await _command.ExecuteReaderAsync();
+                                var resultNumber = 0;
+                                do
+                                {
+                                    while (await reader.ReadAsync())
+                                    {
+                                        _dataReaderAction?.Invoke(reader, resultNumber);
+                                    }
+
+                                    resultNumber++;
+                                } while (await reader.NextResultAsync());
+
+                                break;
+                            }
+                            case ExecType.DataSet:
+                            {
+                                using var adapter = new SqlDataAdapter(_command);
+                                adapter.Fill(result.FillDataSetResult);
+
+                                break;
+                            }
+                            case ExecType.XmlReader:
+                            {
+                                using var reader = await _command.ExecuteXmlReaderAsync();
+                                if (await reader.ReadAsync())
+                                {
+                                    result.XmlResult = await reader.ReadOuterXmlAsync();
+                                }
+
+                                break;
+                            }
+                        }
+
+                        Status = ExecStatus.Success;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    HandleException(ex);
+                }
+            }
+
+            return result;
+        }
+
+        private void HandleException(Exception ex)
+        {
+            Error = ex;
+            Status = ExecStatus.Failed;
+            var handled = _onFailAction?.Invoke(this);
+            if (!handled.HasValue || !handled.Value || !_safe)
```

Comment:
Here is implicit relation between `OnFail `and `Safe` - if `OnFail `action returns `true`(exception was handled) - exception still can be propagated due to the unsafe mode.
Would it be better api something like `Safe(Action onFailAction)` - or something like that (main idea is to have 1 method instead of 2) ?

Reply:
Decided to remove `Safe()` option as not often used. If `Safe` behavior needed, client can use `OnFailure` to prevent exception to be propagated.